src/sentry/tagstore/snuba/backend.py

```diff
@@ -16,69 +17,363 @@
+            'topValues': [{
+                'id': val,
+                'name': tagstore.get_tag_value_label(key, val),
+                'key': tagstore.get_standardized_key(key),
+                'value': val,
+                # TODO we don't know any of these without more queries
+                'count': 0,
+                'lastSeen': 0,
+                'firstSeen': 0,
+            } for val in res['top']],
+        } for key, res in six.iteritems(results)]
 
     # Releases
+    def get_release(self, project_id, group_id, first=True):
+        start, end = self.get_time_range()
+        filters = {
+            'project_id': [project_id],
+        }
+        conditions = [['release', 'IS NOT NULL', None]]
+        if group_id is not None:
+            filters['issue'] = [group_id]
+        aggregations = [['min' if first else 'max', SEEN_COLUMN, 'seen']]
+        orderby = 'seen' if first else '-seen'
+
+        result = snuba.query(start, end, ['release'], conditions, filters,
+                             aggregations, limit=1, orderby=orderby)
+        if not result:
+            return None
+        else:
+            return result.keys()[0]
+
     def get_first_release(self, project_id, group_id):
-        pass
+        return self.get_release(project_id, group_id, True)
 
     def get_last_release(self, project_id, group_id):
-        pass
+        return self.get_release(project_id, group_id, False)
 
     def get_release_tags(self, project_ids, environment_id, versions):
-        pass
+        start, end = self.get_time_range()
+        filters = {
+            'project_id': project_ids,
+            'environment': [environment_id],
+        }
+        # NB we add release as a condition rather than a filter because
+        # this method is already dealing with version strings rather than
+        # release ids which would need to be translated by the snuba util.
+        conditions = [['release', 'IN', versions]]
+        aggregations = [
+            ['count', '', 'count'],
+            ['min', SEEN_COLUMN, 'first_seen'],
+            ['max', SEEN_COLUMN, 'last_seen'],
+        ]
+
+        result = snuba.query(start, end, ['release'], conditions, filters, aggregations)
+
+        return [ObjectWrapper({
+            'times_seen': val['count'],
+            'first_seen': val['first_seen'],
+            'last_seen': val['last_seen'],
+            'key': 'release',
+            'value': name,
+        }) for name, val in six.iteritems(result)]
 
     def get_group_event_ids(self, project_id, group_id, environment_id, tags):
-        pass
+        start, end = self.get_time_range()
+        filters = {
+            'environment': [environment_id],
+            'project_id': [project_id],
+        }
+        # TODO implement environment_id exclusion, its a little bit more complex
+        # than adding a != condition because environment_ids need to be translated
+        # to filters in snuba.
```

Comment:
Would it help if we actually passed in `environment` objects to tagstore methods? Starting to regret the "least knowledge" parameters now. I may see about refactoring that.

Reply:
I think it's more or less ok, with the ids. Its just that the snuba util attempts to translate the columns in `filters` but not `conditions` at the moment. It shouldn't be too hard to add that.